ecc_mock/chaincode/ecc.go

```diff
@@ -52,103 +51,105 @@ func (t *EnclaveChaincode) Invoke(stub shim.ChaincodeStubInterface) pb.Response
-	var errMsg string
-	responseData, signature, errInvoke := t.enclave.Invoke(args, nil, stub)
+	responseBytes, errInvoke := t.enclave.ChaincodeInvoke(stub)
 	if errInvoke != nil {
-		errMsg = fmt.Sprintf("t.enclave.Invoke failed: %s", errInvoke)
+		errMsg := fmt.Sprintf("t.enclave.ChaincodeInvoke failed: %s", errInvoke)
 		logger.Errorf(errMsg)
-		// likely a chaincode error, so we stil want response go back ...
-	}
-
-	fpcResponse := &utils.Response{
-		ResponseData: responseData,
-		Signature:    signature,
-		PublicKey:    []byte("TODO enclave public key"),
-	}
-	responseBytes, _ := json.Marshal(fpcResponse)
-
-	var response pb.Response
-	if errInvoke == nil {
-		response = pb.Response{
-			Status:  shim.OK,
-			Payload: responseBytes,
-			Message: errMsg,
-		}
-	} else {
-		response = pb.Response{
+		// likely a chaincode error, so we still want response go back ...
+		return pb.Response{
 			Status:  shim.ERROR,
 			Payload: responseBytes,
 			Message: errMsg,
 		}
 	}
-	return response
+
+	return shim.Success(responseBytes)
 }
 
 func (t *EnclaveChaincode) endorse(stub shim.ChaincodeStubInterface) pb.Response {
 
-	// TODO
+	chaincodeParams, err := ExtractChaincodeParams(stub)
+	if err != nil {
+		shim.Error(err.Error())
+	}
 
-	// check again chaincode definition and enclave registry
+	responseMsg, err := extractChaincodeResponseMessage(stub)
+	if err != nil {
+		shim.Error(err.Error())
+	}
 
-	// perform enclave signature validation
+	// get corresponding enclave credentials from ercc
+	credentials, err := ercc.QueryEnclaveCredentials(stub, chaincodeParams.ChannelId, chaincodeParams.ChaincodeId, responseMsg.EnclaveId)
+	if err != nil {
+		shim.Error(err.Error())
+	}
+
+	var attestedData protos.AttestedData
+	if err := ptypes.UnmarshalAny(credentials.SerializedAttestedData, &attestedData); err != nil {
+		shim.Error(err.Error())
+	}
+
+	// check cc params match credentials
+	// check cc params chaincode def
+	if ccParamsMatch(attestedData.CcParams, chaincodeParams) {
+		shim.Error("ccParams don't match")
+	}
+
+	// check cc param.MSPID matches MSPID of endorser (Post-MVP)
+
+	// extract read/writes from kvrwset,
+	readset, writeset, err := utils.PerformReadWrites(stub, responseMsg.RwSet)
+	if err != nil {
+		shim.Error(err.Error())
+	}
 
-	// perform reads and writes
-	// for each  readwriteset
-	// 	putState()
-	// 	getState()
+	// Note: below signature was created in ecc_enclave/enclave/enclave.cpp::gen_response
+	// see also replicated verification in tlcc_enclave/enclave/ledger.cpp::int parse_endorser_transaction (for TLCC)
+
+	// H(proposal_payload || proposal_signature || response || read set || write set)
+	hash := utils.ComputedHash(responseMsg, readset, writeset)
+
+	// perform enclave signature validation
+	// TODO refactor this to function
+	pub, err := x509.ParsePKIXPublicKey(attestedData.EnclaveVk)
+	valid := ecdsa.VerifyASN1(pub.(*ecdsa.PublicKey), hash[:], responseMsg.Signature)
+	fmt.Println("signature verified:", valid)
```

Comment:
Regarding outsource of validation following up on comments from yesterday (which disappeared, maybe due to the rebase?), this section plus the Post-MVP comment above is what i would encapsulate in a validation function for `validation.go`.   This is the functionality which would be common also with the plugin way of validating, difference would only be that read/write set and credentials would come via different means?

Reply:
moved this logic to `validation.go`